### DISCUSSION TITLE
Implementing a custom auth flow while taking advantage of RLS

### DISCUSSION THREAD

**Discussion by mstade** [supabase/supabase#544] [Questions]
Hey all!

In our application we're making use of [BankID](https://www.bankid.com/en/) for authenticating our users. The flow looks like so:

1. User is presented with a login screen, and options for how they wish to use their BankID (same device, other device, QR code or personal number.)
2. User picks an option and our application calls the BankID backend to initiate authentication – in OpenID parlance this is akin to making a call to the authorize endpoint to kick of the auth flow, but the difference is that BankID immediately returns a reference code that can eventually be exchanged for a kind of token (see step 6)
3. User is instructed to open their BankID – or, if used on the same device – we open the BankID app for them using a specially crafted BankID URI, e.g. `bankid:///?autostarttoken=[TOKEN]&redirect=[RETURNURL]`
4. User authenticates in the BankID app
5. User comes back to our login screen, if on the same device this is done automatically by way of the `[RETURNURL]` mentioned above
6. Our application polls a BankID endpoint for the status of the authentication using the aforementioned reference code, and if steps 3-5 resulted in a successful authentication then this endpoint will return a kind of token, which is essentially just a JSON blob with some [PII](https://www.investopedia.com/terms/p/personally-identifiable-information-pii.asp) and a signature to be validated against the BankID public key – in OpenID parlance this is akin to calling the token endpoint with an authorization code

At this point we have everything we need to identify a user and log them in. There are no usernames and passwords, so to use this flow with supabase, whenever someone logs in with their BankID we generate a random email for one of our own domains and 128 character random base64 string as password, and use these to sign people up using the API. This is fine and we believe probably reasonably secure, but it raises a few questions.

We can't log people in using usernames and passwords, and there doesn't seem to be an API to generate a JWT to impersonate a given user. Fortunately the JWT secret is given in the settings, so we can presumably use this to generate a token ourselves. But how do we use this with the supabase client?

In #417 there's [a comment](https://github.com/supabase/supabase/discussions/417#discussioncomment-229555) saying you can set the bearer token when creating the client, using initialization options, but [the client docs](https://supabase.io/docs/client/initializing#with-additional-parameters) doesn't have any information about these options. Is the example in the aforementioned comment a stable API, is that the recommended way of passing JWTs we craft ourselves?

I also can't seem to find any documentation on what claims supabase expects the JWT to have. Is there any documentation or guidance on how to craft JWTs in order for them to be accepted by supabase?

Are there any plans to implement OpenID connect or anything like that such that we could actually just use supabase as an OAuth provider backend to our BankID frontend? This would presumable allow for integrations with other auth services like Ping, Okta, Auth0 etc. as well.

Finally, and I know this is getting long (my apologies) but we like to make sure we're getting this right – in the documentation it says you can add user data by using `auth.update()`. It also says you probably shouldn't use this, unless you have data that doesn't change often or pertains specifically to a logged in user. We'll have profile information, application permissions etc. that's definitely only something relevant to the logged in user, but is it recommended we store this stuff in a separate table anyway or should we use the user data option as documented in `auth.update()`? If not the latter, why does this feature exist?

Many thanks for any insights you may be able to provide. I know there's a ton of questions and a lot to read in here, but I'm hoping all this detail might help others seeking the same or similar advice in the future as well. Thank you! 🙏

**Comment by fabianmossberg** [comment]
I'm also very interested in a good implementation of [BankID](https://www.bankid.com/en) and similar solutions such as [Freja eID](https://frejaeid.com/).

What's happened since you wrote this post? Did you find a better workflow?

I'd be happy to contribute to this!

**Comment by matallui** [comment]
Does this help answer your question?
https://github.com/orgs/supabase/discussions/1849
